typescript/trans/typechecking/typescript.nabl2

```diff
@@ -253,7 +255,36 @@ rules
       Return{"function"} <- function_scope,
       Return{"function"} : returnTy,
       Map1 [[ body ^ (function_scope) ]].
-     
+    
+    
+    [[ ClassDeclaration(None(), _, parent, members) ^ (s) ]] :=
+      new clazz,
+      clazz -P-> s,
+      Map1 [[ members ^ (clazz) ]],
+      ClassParent [[ parent ^ (s): parentTy ]].
+      
+    [[ ClassDeclaration(Some(name), _, parent, members) ^ (s) ]] :=
+      new clazz,
+      clazz -P-> s,
+      Value{name} <- s,
+      classTy == RECORD(clazz),
+      Value{name} : classTy !,
+      Type{name} <- s,
+      Type{name} : classTy !,
+      Value{"this"} <- clazz,
+      Value{"this"} : classTy !,
+      ClassParent [[ parent ^ (s): parentTy ]],
+      classTy <! parentTy,
+      Map1 [[ members ^ (clazz) ]].
+    
+    ClassParent [[ None() ^ (s): RECORD(s) ]] := true.
+    ClassParent [[ Some(ClassExtends(parent)) ^ (s): parentTy ]] :=
+       [[ parent ^ (s): parentTy ]],
+       parentTy == RECORD(parentScope) | error $[Expected class as parent, but got [ty]],
+       Value{"this"} -> parentScope,
```

Comment:
This line fails and I have no clue why. For some reason it can not find anything in this `parentScope`. The `parentScope` is defined on line 270 and I don't understand why it would not find it. Nabl2 spews the following errors which I don't understand:
```
Unsolved: distinct/name Declarations(#-program_scope-1,Namespace("Type"))
Unsolved: distinct/name Declarations(#-program_scope-1,Namespace("Value"))
Unsolved: RECORD(#-clazz-2) <defaultrelation()! record(?-parentscope-1)
Unsolved: Value{"this" @:14} -> ?-parentScope-1
Expected a class, but got ?-parentScope-1
Unsolved: ?-d-2.Type() := RECORD(?-parentScope-1)
Cannot resolve "Bar"

Type: RECORD(?-parentScope-1)
```

Also this is kinda a hack, because on a property lookup we don't know whether we are dealing with a record or a class. Since you can't encode an or with the constraints, we have to mimic classes as records. Thus to distinguish records with classes, I tried to look up the `Value{this}`, because only classes should have them. This is a complete hack, but I don't know a different method :(